### PR TITLE
Stockfish jr

### PR DESCRIPTION
**Pull request by jrosengarden**
I've always wished for some "help" ability within Stockfish at the command line.
With the release of Stockfish v15 I decided to finally add this ability.

This pull request ONLY adds a light-weight help system to StockFish's command line.  
To access it simply type "help" at any time.

There are no operational changes at all....just the lightweight help system.

I'd really love to see this added into the master branch.......I think many, many, people would enjoy/appreciate the help feature(s).

### CODE REVIEW

**Comment by dubslow** [comment]
in principle, having a "help" system will go a long way. took me ten minutes to figure out how to use stockfish the first time, couldn't believe there was no "help" at the time.

that said, putting your name in all those comments is more than a bit weird, also the brace formatting should match the rest of the code base

**Comment by Karma-Tron** [comment]
This feature makes the assumption that everyone is using 24 line command windows. Personally I would leave that out.

**Comment by jrosengarden** [comment]
> in principle, having a "help" system will go a long way. took me ten minutes to figure out how to use stockfish the first time, couldn't believe there was no "help" at the time. even the next few times i had to continually re-google UCI, having a quick reference in the program would aid utility a lot.
> 
> that said, putting your name in all those comments is more than a bit weird, also the brace formatting should match the rest of the code base

I put all the comments in so I could find the code changes easily.  I figured I'd remove them if the change was accepted for merging with the master.

**Comment by jrosengarden** [comment]
> Nice feature for those who were looking for this. However, it makes the assumption that everyone is using 24 line command windows. Personally I would leave that out.
> Same for `(end) mod by **JR** on 5/3/22`. Nobody comments a change with their name I think. Could become quite unreadable for nesting changes.

I put all the comments in so I could find the code changes easily. I figured I'd remove them if the change was accepted for merging with the master.

**Comment by jrosengarden** [comment]
> Nice feature for those who were looking for this. However, it makes the assumption that everyone is using 24 line command windows. Personally I would leave that out.
> Same for `(end) mod by **JR** on 5/3/22`. Nobody comments a change with their name I think. Could become quite unreadable for nesting changes.

I put all the comments in so I could find the code changes easily. I figured I'd remove them if the change was accepted for merging with the master.

**Comment by dubslow** [comment]
Well it certainly *won't* be accepted *until* those comments are changed. The maintainer certainly isn't going to remove the comments themself

**Comment by Karma-Tron** [comment]
> Nice feature for those who were looking for this. However, it makes the assumption that everyone is using 24 line command windows. Personally I would leave that out.

I should add that I meant the clrscr function.

**Comment by vdbergh** [comment]
It seems this simply duplicates the description of the UCI protocol.

Note: stockfish is meant to be used via a GUI that speaks the UCI protocol. It is therefore assumed that people that invoke stockfish via the command line are developers that know what they are doing,

**Comment by jrosengarden** [comment]
> It seems this simply duplicates the description of the UCI protocol.
> 
> Note: stockfish is meant to be used via a GUI that speaks the UCI protocol. It is therefore assumed that people that invoke stockfish via the command line are developers that know what they are doing,

Yes it does "duplicate" the description(s) of the UCI protocol...but it puts them at the developers fingertips - within the program - for ez reference (something I've wished was there hundreds of times).

Yes - stockfish is meant to be used with a GUI but the command line ability is there for anybody to use for whatever purpose they may need to be at the command line so....again.....giving them a quick reference within the program is nothing but helpful.

**Comment by jrosengarden** [comment]
To All - Regarding Comments:

I have removed all the **JR** comments and replaced them with "HELP System".  This will allow somebody, in the future, to hunt down any code related to this lightweight help system.

**Comment by jrosengarden** [comment]
> > Nice feature for those who were looking for this. However, it makes the assumption that everyone is using 24 line command windows. Personally I would leave that out.
> 
> I should add that I meant the clrscr function with the remark above. Original post edited as well.

I agree that others might be at the command line with a terminal bigger/smaller than 24 lines.  However I don't want to remove the clrscr() ability since it "cleans" the screen before displaying the Help System menu (and any of the detailed help screens).  I could increase the "clearing" function for a larger terminal (i.e. - clear 48 lines instead of 24).  However for smaller screens "clearing" 24 lines still accomplishes the desire to have a "clean slate" when presenting the information.

**Comment by pschneider1968** [comment]
Please don't clear the screen/terminal buffer through this help command. You will surely erase information that I still might want to refer to, since I'm working in the CLI, e.g. other engine output, bench number, whatever. Just write the HELP information to STDOUT without clearing the screen.

**Comment by jrosengarden** [comment]
To All: Regarding the ClrScr() function:
Based on popular demand I will pull the ClrScr() function and NOT clear the screen when displaying the Help System info.  IMHO that tends to wind up with a cluttered screen but Peter Schneider made a good case for not doing so in case the user has some info already on the screen and doesn't want it erased.  Once the current checks complete I'll update the pull request with this change.

Thanks for the input everybody.

**Comment by nickolasreynolds** [comment]
If a help command is absolutely necessary, it seems to me that it would make more sense to simply output a URL with an appropriate link, e.g. https://stockfishchess.org/help (which doesn't currently exist.)

**Comment by vondele** [comment]
I also don't think this is the right path. It is mostly a duplication of documentation that we already have https://github.com/official-stockfish/Stockfish#the-uci-protocol-and-available-options I would prefer we work on the more visible documentation (like wiki or web pages) rather than doing this at the engine level.

**Comment by dubslow** [comment]
@vondele rest assured that my experience when I first tried to "into stockfish" was nowhere near as easy as you seem to think. The README does a good job of enumerating uci options, but it is *nowhere* near sufficient to figure out how to use the binary. What is necessary to use  the binary is in fact having the wherewithal to google up something like http://wbec-ridderkerk.nl/html/UCIProtocol.html , and spend 10 minutes reading and slowly stumbling your way into using uci commands and figuring how the heck the README is supposed to be useful.

At the *very* very minimum, a nice `--help` message (with both cmdline option `--help` or with the command `help`, as in this PR), with a 3 sentence explanation of UCI, a link, and a list of basic commands, which will go a long way towards making the README useful -- since, as I said, it isn't remotely useful to a beginner right now. Perhaps you can argue that having the entire uci specification printed to terminal is a bit excessive, but frankly I would love to have that too as it's quite a lot more convenient than having to google up the link I pasted above.

Long story short, `stockfish` is one of the least beginner-friendly binaries I've ever come across, and that's really saying something considering some of the crap that can be found in the wider linux/open source world.

PRs like this should be wholeheartedly embraced and welcomed, and pending the usual code quality checks, should be pushed thru to completion promptly. (If you perhaps want to start with a simpler help message, as I and others have postulated, fine start there, but the fuller version in this PR is still a net benefit IMO.)

**Comment by nickolasreynolds** [comment]
> PRs like this should be wholeheartedly embraced and welcomed, and pending the usual code quality checks, should be pushed thru to completion promptly.

Ah, yes, here we have what is surely one of the most potent persuasion techniques, simply telling people that they should agree with you, and _immediately_. No doubt the Stockfish project will continue to languish in obscurity without your seminal contributions!

**Comment by Sopel97** [comment]
@dubslow 

why are you even using stockfish through cli

**Comment by ppigazzini** [comment]
> The README does a good job of enumerating uci options, but it is _nowhere_ near sufficient to figure out how to use the binary.

Improving the README, or better writing a new wiki page, will make easy to read/share the information.

**Comment by jrosengarden** [comment]
Couple of comments:
1.  I'm not sure what the REAL basis of the objections/arguments for NOT including a lightweight help system within the engine are.
     a.- it does not HURT/HINDER the engine, or it's performance, AT ALL.
     b.-it goes a long way to providing a friendlier environment WITHIN the engine.
     c.-yes there are web pages out there that one can turn to/refer to...but (for a newbie) only after much hunting around.
          Having the lightweight help system WITHIN the engine will HELP them tremendously.
     d.- can it be made even better?  YOU BET.  Once it's merged into the master I'm sure dozens of bright programmers
           will provide PRs that improve it tremendously.  THIS PR is simply a 'starting ground' for something that IMHO
           should have always been there.

2.  Could it be that, as a long time experienced Stockfish contributor you are only "seeing" this from YOUR perspective?
     Of course YOU don't need a help system.  Of course there are web sites out there that YOU are familiar with.
     Of course this PR is simply regurgitating info that YOU are already knowledgeable about.
     However......remind yourself what a help system is for:  To provide basic (or in-depth) assistance to a software user who
    IS NOT as familiar with the world of info available on the topic as the software developer/contributors are.
    Just pointing to other sources/locations of available information, as THE argument for not including a help system,  is 
    counter to the idea AND usefulness of an in-app help systems.

   Which brings to mind my final (somewhat passive/aggressive) comment:

I would suggest that those of you objecting to the inclusion of a lightweight help system within Stockfish's engine examine your premise for objecting!!

NOTE:  Based on a couple of very useful comments this morning's commit adds a pointer to a web site that contains a very good description of the UCI protocol (that many of you have indicated you are familiar with).  Adding this web url to the help system was a good idea....giving that newbie (or even an experienced developer/contributor who, like me, forgets things)......quick access to a starting point on the web.

**Comment by Sopel97** [comment]
Change help output to "download banksia" and I'll be fine with it. If someone needs help with how to use UCI they are already doing things wrong when they need to type help.

**Comment by jrosengarden** [comment]
> Change help output to "download banksia" and I'll be fine with it. If someone needs help with how to use UCI they are already doing things wrong when they need to type help. It is not an intention of this project to reiterate UCI spec in 10 places.

Not sure what you mean by "Change help output to "download banksia"".  Please be more specific.   Thanks!

**Comment by Sopel97** [comment]
If you don't know UCI you're not supposed to run stockfish itself at all. It's like if GCC had --help that teaches you C.

**Comment by ppigazzini** [comment]
IMO the [wiki](https://github.com/glinscott/fishtest/wiki) is the natural place where write information. There is the possibility to add examples, little scripts, links to external sites etc. It's easy for all the contributors to review/improve the old content and add new content. I started contributing to the wiki as a newbie adding all the new things learned to help other newbies.
by thw way fishtest worker has a cli help showing the options but I wrote a wiki section too adding some extra info.

**Comment by Sopel97** [comment]
> > If you don't know UCI you're not supposed to run stockfish itself at all. It's like if GCC had --help that teaches you C.
> 
> Ahhh....I get it. So under that logic anybody that doesn't know how to utilize (for example) Excel shouldn't be using it....OR MAYBE that's why there is an internal help system within (for example) Excel for both those that don't know too much about how it works or, even better, for those of us that constantly forget deeper details about Excel and need a quick reminder...without having to go run around the WWW trying to find that information.

Excel is meant to be used manually by non-developer people. Stockfish not meant to be used like that.

**Comment by pschneider1968** [comment]
I agree very strongly to the idea of adding a help command to Stockfish. As it is now without a help command, using Stockfish from the command line is not very approachable. It is not only developers who would like to use SF from the CLI, but chess + computer chess enthusiasts who want to take a deep dive into what is possible with state of the art chess engines.

Also, as a matter of course, any command line tool should have a good help command. Imagine the GNU tools without help command! How would you be supposed to learn them?

**Comment by pschneider1968** [comment]
> If you don't know UCI you're not supposed to run stockfish itself at all. It's like if GCC had --help that teaches you C.

GCC has an extensive help command that clarifies the use of all the available command line options. It is a quick reference for someone already familiar with the tool, but may have forgotten about a specific option. In the same way, a help command for SF would be a handy quick reference for someone who is already somewhat familiar with UCI, but cannot memorize all of it when using the SF command line.

**Comment by pschneider1968** [comment]
> 
> PRs like this should be wholeheartedly embraced and welcomed, and pending the usual code quality checks, should be pushed thru to completion promptly. (If you perhaps want to start with a simpler help message, as I and others have postulated, fine start there, but the fuller quick reference in this PR is still a net benefit IMO.)

I'll second that!

**Comment by jrosengarden** [comment]
Hi All:
With this latest commit (moments ago) there are now three choices:
1.  Give stockfish the internal engine help system that we've been arguing about for the last couple of days.
2. Give stockfish a command line interface help system (that I just committed)
    -  simply type ./stockfish --help       (the rest is self-evident)
3.  Give stockfish BOTH the internal engine help system AND the command line interface help system

NOTE:  I was slightly surprised as I was getting ready to push the new command line interface.  I was going to comment out all of the code for the internal engine help system...so nobody would be confused.....but it turns out they work hand-in-hand...with a feature I hadn't even figured out earlier.  

Right now the command line interface starts with ./stockfish --help   [which brings up the external command line help]
---but you could also do this:  ./stockfish help    [which brings up the internal engine help system]
Right now, with the command line interface, you can do this:  ./stockfish --help ucinewgame
      -- which will bring up the external command line interface specific help regarding the ucinewgame command.
--but you could also do this:  ./stockfish ucinewgame/help
     -- which will bring up the internal help system regarding the ucinewgame command.

I am in NO WAY suggesting that we leave BOTH of them in.  However...at the moment.....both are working just fine.  I'll need to hear from the 'gallery' to decide which one to leave in and which one to cut.....or surprise:surprise....leave 'em both there.

**Comment by pschneider1968** [comment]
I'd vote for option 3 - keep them both. Help should be as easily accessible as possible, IMHO, and redundant ways to get essential information make it easier.

**Comment by jrosengarden** [comment]
To All:
I've added one more commit today (and I promise this will be my last for the day).
This commit brings the external CLI help system in-line with the SF internal help system.
They now both present identical menus, identical text, etc.
The SF internal help menu no longer has asterisks next to the commands that had "add'l" help.  They now ALL have add'l help - again...to match the external CLI help system interface.

I really do like the idea of having both help systems in place.  The "back each other up".

OK...that's it for today!   :-)

**Comment by Sopel97** [comment]
As a compromise I suggest adding output for --help cli switch that suggest using some specific existing GUI and links to the UCI spec

**Comment by jrosengarden** [comment]
> As a compromise I suggest adding output for --help cli switch that suggest using some specific existing GUI and links to the UCI spec

I'm not sure if the base SF binary wants to be recommending a particular GUI (now THAT would make a whole lot of folks see red)!!

I have DEFINITELY added an URL to the UCI Protocol page that others have suggested (above).  This URL appears on both the SF CLI help system interface AND the SF Internal help system interface.

**Comment by MichaelB7** [comment]
@vondele has once again provided the most practical guidence. The exe itself should not be burdened with a full blown help system.  That was fine 30 years ago when other options were more limited or simply not available.  It would be nice if anyone has the time to work o this on Wiki or on web display. The visual enhancements with screenshots , structure and formatting would be a major improvement over what we have today and what is proposed here.

**Comment by ppigazzini** [comment]
Lazy dump of the menu from this PR, to be improved by the community (I'm not an expert in UCI commands).
https://github.com/glinscott/fishtest/wiki/Stockfish%3A-the-missing-menu

**Comment by ddugovic** [comment]
There is technical documentation at the developer's fingertips: `man stockfish`.  (Arguably this could be updated or link to GitHub.)

```
STOCKFISH(6)                                                      STOCKFISH(6)

NAME
       stockfish - free UCI chess engine, to calculate chess moves

SYNOPSIS
       stockfish

DESCRIPTION
       This manual page documents the stockfish command.

       stockfish is an UCI chess engine, to calculate chess moves

       Stockfish  is  a free UCI chess engine derived from Glaurung 2.1. It is
       not a complete chess program, but  requires  some  UCI  compatible  GUI
       (like  XBoard  with  PolyGlot, eboard, Jose, Arena, in order to be used
       comfortably.  Read the documentation for your GUI of choice for  infor‐
       mation about how to use Stockfish with your GUI.

       This  version  of  Stockfish  supports  up  to 8 CPUs, but has not been
       tested thoroughly with more than 2.  The program tries  to  detect  the
       number  of  CPUs  on your computer and set the number of search threads
       accordingly, but please be aware that the detection is not always  cor‐
       rect.   It  is  therefore  recommended  to  inspect  the  value  of the
       "Threads" UCI parameter, and to make sure it equals the number  of  CPU
       cores  on your computer. If you are using more than four threads, it is
       recommended to raise the value of "Minimum Split Depth"  UCI  parameter
       to 6.

       Opening book

       This version of stockfish has experimental support for PolyGlot opening
       books.  For information about how to create  such  books,  consult  the
       PolyGlot documentation.

OPTIONS
       This  program  does  NOT follow the usual GNU command line syntax, with
       long options starting with two dashes.

SEE ALSO
       Communication with UCI commands is documented by "The UCI Communication
       Protocol" in /usr/share/doc/stockfish/engine-interface.txt.

AUTHOR
       stockfish  was  written  by <Romstad Tord <tord@glaurungchess.com>> and
       Marco Costalba <mcostalba@gmail.com>.

       This manual page was written by Oliver Korff <ok@xynyx.de>, for the De‐
       bian project (but may be used by others).

                               October  10, 2009                  STOCKFISH(6)
```

**Comment by vondele** [comment]
That man page is actually part of the debian project https://manpages.debian.org/testing/stockfish/stockfish.6.en.html (and indeed somewhat outdated)

But indeed, if we would properly install a stockfish package, such a man page could be part of the package. Maybe even automatically generated from the README.md ? (There seemingly is a tool for that https://github.com/eush77/readme-to-man-page, untested). 

Our official downloads do contain the sources and the corresponding readme, abrok links to the sources, IIRC.

**Comment by Sopel97** [comment]
a .chm file could be generated for windows too

**Comment by vondele** [comment]
I'm considering something like this:

```diff
diff --git a/src/uci.cpp b/src/uci.cpp
index 7b30cc043..74affda79 100644
--- a/src/uci.cpp
+++ b/src/uci.cpp
@@ -285,8 +285,13 @@ void UCI::loop(int argc, char* argv[]) {
               filename = f;
           Eval::NNUE::save_eval(filename);
       }
+      else if (token == "--help" || token == "help" || token == "--license" || token == "license")
+          sync_cout << "\nStockfish is a free, powerful chess engine, provided under the GNU General Public License version 3."
+                       "\nStockfish is a command line program that requires a separate graphical user interface (GUI) to be used comfortably."
+                       "\nYou should have received a copy of the license (Copying.txt) and the documentation (README.md) with this program."
+                       "\nThese files can also be found at https://github.com/official-stockfish/Stockfish\n" << sync_endl;
       else if (!token.empty() && token[0] != '#')
-          sync_cout << "Unknown command: " << cmd << sync_endl;
+          sync_cout << "Unknown command: '" << cmd << "'. Type help for more information." << sync_endl;
 
   } while (token != "quit" && argc == 1); // Command line args are one-shot
 }
```
this would print the small text when `--help` or `--license` is added to command line, or `help` or `license` is typed as a command.

Opinions?

**Comment by Disservin** [comment]
I would add something like
`A collection of commands can be found at gitub.com/official-sotckfish/etc., or type uci`

**Comment by vondele** [comment]
note that we don't list the uci commands, just the options.

**Comment by Disservin** [comment]
Noted, but the command is help, and `These files can also be found at https://github.com/official-stockfish/Stockfish` is not very helping in my opinion ^^

**Comment by vondele** [comment]
let's look for alternatives... not duplicating information

**Comment by vondele** [comment]
I'll close this PR as it won't be merged in this form. However I have created #4027 as an alternative.
